src/historywork/WriteVerifiedCheckpointHashesWork.cpp

```diff
@@ -0,0 +1,169 @@
+// Copyright 2020 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "WriteVerifiedCheckpointHashesWork.h"
+#include "catchup/CatchupConfiguration.h"
+#include "catchup/CatchupRange.h"
+#include "catchup/DownloadApplyTxsWork.h"
+#include "catchup/VerifyLedgerChainWork.h"
+#include "history/FileTransferInfo.h"
+#include "history/HistoryManager.h"
+#include "historywork/BatchDownloadWork.h"
+#include "ledger/LedgerManager.h"
+#include "ledger/LedgerRange.h"
+#include "main/Application.h"
+#include "util/Logging.h"
+#include "work/ConditionalWork.h"
+#include <Tracy.hpp>
+#include <fmt/format.h>
+
+namespace stellar
+{
+
+class TmpDirOwnerWork : public Work
+{
+    TmpDir const mTmpDir;
+
+  public:
+    TmpDirOwnerWork(Application& app, std::string const& name)
+        : Work(app, name, RETRY_NEVER)
+        , mTmpDir(mApp.getTmpDirManager().tmpDir(getName()))
+    {
+    }
+    BasicWork::State
+    doWork() override
+    {
+        return State::WORK_SUCCESS;
+    }
+    TmpDir const&
+    getTmpDir()
+    {
+        return mTmpDir;
+    }
+};
+
+WriteVerifiedCheckpointHashesWork::WriteVerifiedCheckpointHashesWork(
+    Application& app, LedgerNumHashPair rangeEnd, std::string const& outputFile,
+    std::shared_ptr<HistoryArchive> archive)
+    : BatchWork(app, "write-verified-checkpoint-hashes")
+    , mRangeEnd(rangeEnd)
+    , mCurrCheckpoint(rangeEnd.first)
+    , mArchive(archive)
+    , mOutputFile(std::make_shared<std::ofstream>(outputFile))
+{
+    if (mArchive)
+    {
+        CLOG(INFO, "History")
+            << "WriteVerifiedCheckpointHashesWork: selected archive "
+            << mArchive->getName();
+    }
+    if (!*mOutputFile)
```

Comment:
Maybe I'm not reading the [spec](http://www.cplusplus.com/reference/ios/ios/exceptions/) incorrectly, but I think we should set `badbit` on that stream that way any io error will properly throw and we don't need to check for errors all over the place (which we don't btw; and it looks like there are other places in the code where we don't set this either - like when we write the has file).

Reply:
I'm not sure I understand -- we are detecting badbit and throwing on error here. Setting badbit again doesn't make sense to me.

Reply:
yes we do here, what I was saying is that we don't check for failures in subsequent writes

Reply:
Ah! You're talking about setting the stream's _exception mask_, sorry, I misread. For sure, makes sense (I had forgotten this facility even existed :P)

It's a good point about not setting this everywhere we should. I'll open another PR shortly.

Reply:
Yeah it's one of those non-sense "unsafe by default" that the standard went with. Thanks